[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107951
* #106958
* #106838
* #106836
* #107565
* #105996
* #105906
* __->__ #105639
* #105456
* #105455
* #104590
* #104588
* #104581
* #104580

**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```

leslie-fang-intel committed:
[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_non_quantizable_op_after_force_int8_int8_op
```


[ghstack-poisoned]

jgong5 commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -369,6 +472,124 @@ def _annotate_conv2d(
+        # Recipe refers to https://github.com/intel/intel-extension-for-pytorch/blob/
+        # 90d19323d96afc53fcc22ba5a7bb3fb07fdd6c1c/intel_extension_for_pytorch/quantization/_utils.py#L495
+        r"""
+        1. Force insert observer at output of this node.
+        2. It also fixes this case:
+        For ops which we want to force int8 in and int8 output, there is a case when the op has multi user nodes.
+        If some of the user nodes are not quantizable, we need to lift up obsever from user node's input into output.
+        Otherwise fail to find can't find dq - force_int8_node - quant pattern.
+        # Before
+             observer
+                |
+              view
+              /   \
+        observer non-quantized_op
+           /
+        quantized_op
+        # After
+                observer
+                  |
+                 view
+                  |
+                observer
+                /   \
+        quantized_op non-quantized_op
```
May I know how this comment is related to the implementation here?

leslie-fang-intel replied:
We can remove this comment. It's just about another reason why we need to insert observer at output of some node. Because, it can lift up the position of fake quant. Then, we can find the quantization pattern to match and lowering.

jgong5 commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -369,6 +472,124 @@ def _annotate_conv2d(
+    def _annotate_maxpool2d(
+        self, node: Node, quantization_config: QuantizationConfig
+    ) -> None:
+        if (
+            node.target is not torch.ops.aten.max_pool2d_with_indices.default
+            or not (
+                len(list(node.users)) == 1
+                and (list(node.users)[0].target == operator.getitem)
+            )
+        ):
```
If it is a pattern, why not handle it like conv fusion but treat it as a single op?

leslie-fang-intel replied:
Thanks for the comment. After second thought, I think `_annotate_quantizable_single_op` may not be a good name here which causes confusion. And I think we should not annotate `maxpool2d` like `conv fusion`. Here is some of my understanding:

- The difference between `conv` and `maxpool2d` is not a pattern or single op.
- `Conv` should be more like the `lower_precision_fp` op set in `Autocast`, which we will force to run with lower precision to get better performance.
- `Maxpool2d` is more like the fall-through op set in `Autocast`, which will run with lower precision if previous node does so.

So, how about we change the function name from `_annotate_quantizable_single_op` to something may be like `_annotation_propagation_quantizable_pattern`? @jgong5 do you have any suggestions here?

leslie-fang-intel replied:
Then, maybe we can abandon the concept of `single quantized node` to unify `Conv` and `Maxpool2d` as `quantized pattern`.

jgong5 commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -245,16 +323,38 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+            # in inputs. So, we can fuse dq-operator-q into a quantized op.
+            # Refer to https://github.com/intel/intel-extension-for-pytorch/blob/
+            # 90d19323d96afc53fcc22ba5a7bb3fb07fdd6c1c/intel_extension_for_pytorch/quantization/_recipe.py#L487
+            self._annotate_quantized_node_output(node, config)
```
It actually handles `int8_in_int8_out_ops_pt2e` right? If so, why not name it that way?

leslie-fang-intel replied:
Thanks for the commend and changed.

jgong5 commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -28,18 +29,94 @@
+}
+
+# Ops input and output share observer.
+in_out_share_obs_ops_pt2e: Set = {
```
Where is it used?

leslie-fang-intel replied:
It's not used yet. Remove it in this PR.

jgong5 commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -245,16 +323,38 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
     def _annotate_for_static_quantization_config(
         self, model: torch.fx.GraphModule
     ) -> torch.fx.GraphModule:
```
can we have a high-level description of the quantization recipe here?

leslie-fang-intel replied:
Thanks for the comment. The high-level description of the quantization recipe is added here.

jgong5 replied:
Nice description. Thanks!

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -28,18 +29,94 @@
+    check if any of the node is annotated, return True if any of the node
+    is annotated, otherwise return False
+    """
+    return (len(nodes) != 0) and any(_is_node_annotated(node) for node in nodes)
```
Thanks for the comments. Remove the check of `(len(nodes) != 0)`.

leslie-fang-intel commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -28,18 +29,94 @@
+    Given a list of nodes (that represents an operator pattern),
+    return True if all of the node is annotated, otherwise return False
+    """
+    return (len(nodes) != 0) and all(_is_node_annotated(node) for node in nodes)
```
Thanks for the comment. I haven't thought about it deeply in previous. I think we can align the behavior of `_is_all_annotated` with `all` in python. So, return `True` of `_is_all_annotated` when empty list passed in. So, removed the check of `(len(nodes) != 0)`.

leslie-fang-intel commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -28,18 +29,94 @@
+    Case2: the node has been annotated as single quantized node.
+    """
+    quantization_annotation = node.meta.get("quantization_annotation", None)
+    if (not _is_any_annotated([node])) or (not quantization_annotation):
```
Remove the redundant check here.

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

jgong5 commented on torch/ao/quantization/pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -84,37 +76,33 @@ def _is_any_annotated(nodes: List[Node]):
-    return False
+    quantization_annotation = node.meta.get("quantization_annotation", None)
+    assert isinstance(quantization_annotation, _X86InductorQuantizationAnnotation)
+    return True if (
+        quantization_annotation._is_output_of_quantized_pattern
+    ) else False
```
```suggestion
    return quantization_annotation._is_output_of_quantized_pattern
```

leslie-fang-intel replied:
Oh, thanks and changed.

jgong5 reviewed (approved):
Overall, LGTM. Please check minor suggested edits.

leslie-fang-intel commented:
Hi @jerryzh168 @kimishpatel, could you kindly help to take a look of this PR?

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168, could you help to take a look of this PR?

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -28,14 +29,83 @@
     SourcePartition,
 )
-from .quantizer import QuantizationAnnotation, QuantizationSpec, Quantizer
+from .quantizer import (
```
nit: probably better to import from "torch.ao.quantization.quantizer"

leslie-fang-intel replied:
Thanks for comment and changed.

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -389,6 +506,104 @@ def _annotate_conv2d(
+                    list(maxpool_node.users)[0].target == operator.getitem
+                )
+                getitem_node = list(node.users)[0]
+                if not _is_all_annotated([getitem_node, maxpool_node]):
```
you probably want to skip if any of the node is annotated? otherwise you will be breaking some existing annotations?

leslie-fang-intel replied:
Oh, I think here we add this check because we expected `getitem_node` and `maxpool_node` already been annotated in previous step of `_annotation_propagation_quantizable_pattern` which should only annotate the inputs of these 2 nodes. Then, here we will annotate the output of `getitem_node` here.

jerryzh168 replied:
I see, I feel this is error prone, since these might be annotated because of other reasons, I think the more robust way to annotate things would be to have a self contained meaningful pattern and just annotate that pattern by itself.

leslie-fang-intel replied:
Make sense. We may discuss about it further.

jerryzh168 reviewed (approved):
accepting to unblock

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d)"


**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```


[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
